scalameta/contrib/shared/src/main/scala/scala/meta/contrib/ScaladocTokenizer.scala

```diff
@@ -0,0 +1,184 @@
+  }
+
+  private val trailTextParser: Parser[Text] = P(nl0Hspaces0 ~ textParser)
+  private val leadTextParser: Parser[Text] = P(leadHspaces0 ~ textParser)
+
+  private val tagParser: Parser[Tag] = {
+    val tagTypeMap = TagType.predefined.map(x => x.tag -> x).toMap
+    def getParserByTag(tag: String): Parser[Tag] = {
+      val tagTypeOpt = tagTypeMap.get(tag)
+      tagTypeOpt.fold[Parser[Tag]] {
+        trailTextParser.map { desc => Tag(TagType.UnknownTag(tag), desc = desc) }
+      } { tagType =>
+        (tagType.hasLabel, tagType.hasDesc) match {
+          case (false, false) => PassWith(Tag(tagType))
+          case (false, true) => trailTextParser.map(x => Tag(tagType, desc = x))
+          case (true, false) => trailWordParser.map(x => Tag(tagType, label = x))
+          case (true, true) =>
+            (trailWordParser ~ trailTextParser).map {
+              case (label, desc) => Tag(tagType, label, desc)
+            }
+        }
+      }
+    }
+    val tagLabelParser = P(("@" ~/ labelParser).!)
+    P(leadHspaces0 ~ tagLabelParser.flatMap(getParserByTag))
+  }
+
+  private def listBlockParser(minIndent: Int): Parser[ListBlock] = {
+    val listParser = (hspacesMinWithLen(minIndent) ~ listPrefix.! ~ hspaces0).flatMap {
+      case (indent, prefix) =>
+        val sep = (nl ~ hspacesMinWithLen(indent) ~ prefix).flatMap { x =>
+          if (x != indent) Fail else Pass
+        }
+        (textParser ~ listBlockParser(indent + 1).?)
+          .map { case (desc, list) => ListItem(desc, list) }
+          .rep(1, sep = sep)
+          .map(x => ListBlock(prefix, x.toSeq: _*))
+    }
+    P(startOrNl ~ listParser)
+  }
+
+  /** Contains all scaladoc parsers */
+  private val parser: Parser[collection.Seq[ScaladocToken]] = {
+    val allParsers = Seq(
+      listBlockParser(1),
+      codeBlockParser,
+      headingParser,
+      tagParser,
+      leadTextParser // keep at the end, this is the fallback
+    )
+    P(allParsers.reduce(_ | _).rep(1) ~/ End)
+  }
+
+  private val scaladocLine = Pattern.compile("^[ \t]*\\**(.*?)[ \t]*$")
+
+  /** Parses a scaladoc comment */
+  def tokenize(comment: Comment): Option[Seq[ScaladocToken]] =
+    tokenize(comment.syntax)
+
+  /** Parses a scaladoc comment */
+  def tokenize(comment: String): Option[Seq[ScaladocToken]] =
+    CommentOps.getIfScaladoc(comment).map(tokenizeImpl)
+
+  private def tokenizeImpl(comment: String): Seq[ScaladocToken] = {
+    val sb = new java.lang.StringBuilder
+    val res = Seq.newBuilder[ScaladocToken]
+    def flush: Boolean =
+      sb.length() != 0 && {
+        val paragraph = sb.toString
+        sb.setLength(0)
+        parser.parse(paragraph) match {
+          case p: Parsed.Success[collection.Seq[ScaladocToken]] =>
+            res ++= p.value
+            p.value.nonEmpty
+          case _ =>
+            res += Unknown(paragraph)
+            true
+        }
+      }
+
+    var flushed = false
+    comment.linesIterator.foreach { line =>
```

Comment:
one other comment you made in that old pr is to avoid text manipulation outside of fastparse. while i did that, compared to the previous implementation, by combining all parsers into one, there was one thing that i couldn't get fastparse to do.

i couldn't write a parser that would be able to identify paragraph breaks (one or more black lines) and at the same time not consume the last newline. keeping that newline was critical to implementing the list block parser, which is sensitive to indentation at the beginning of a line.

so i decided to split code into paragraphs, stripping leading space before asterisk and trailing space, and applying fastparse to the result.